### PR TITLE
Upgrade chem/aeronet aod and aaod converter to IODA V2 API

### PR DESCRIPTION
**Pull request by bhuang95**
## Description
IODA converter for aeronet aod and aaod is upgraded to IODA V2 API.
Note AttrData['nchans'] is currently not correctly assigned yet (see Dependencies). 
Output IODA V2 files show similar variable values except for "nchan" variable compared to the files generated through ioda-upgrader.x

[See the JEDI Documentation](https://jointcenterforsatellitedataassimilation-jedi-docs.readthedocs-hosted.com/en/latest/inside/practices/pullrequest.html) for further information.

### Issue(s) addressed
This PR is related to issue #654  .
(https://jointcenterforsatellitedataassimilation-jedi-docs.readthedocs-hosted.com/en/latest/inside/practices/issues.html)

## Acceptance Criteria (Definition of Done)

What does it mean for this PR to be finished?

## Dependencies
Waiting for another PR JCSDA-internal/ioda/pull/508 to correct dimension index. Then these python codes will be further changed to correctly assign AttrData['nchans']. 

## Impact

Please list the other repositories (if any) that this PR will require changes in (example below)

- [ ] saber
- [ ] ufo
- [ ] ...

### CODE REVIEW

File: src/chem/aeronet_aod2ioda.py [hunk 1]
```diff
@@ -104,76 +97,92 @@ def add_data(infile):
...
     for key, value in obsvars.items():
-        outdata[varDict[key]['valKey']] = np.array(f3[value].fillna(nc.default_fillvals['f4']))
-        outdata[varDict[key]['qcKey']] = np.where(outdata[varDict[key]['valKey']] == nc.default_fillvals['f4'],
+        outdata[varDict[key]['valKey']] = np.array(f3[value].fillna(np.float32(-999.)))
+        outdata[varDict[key]['qcKey']] = np.where(outdata[varDict[key]['valKey']] == np.float32(-999.),
                                                   1, 0)
-        outdata[varDict[key]['errKey']] = np.where(outdata[varDict[key]['valKey']] == nc.default_fillvals['f4'],
-                                                   nc.default_fillvals['f4'], 0.02)
+        outdata[varDict[key]['errKey']] = np.where(outdata[varDict[key]['valKey']] == np.float32(-999.),
+                                                   np.float32(-999.), 0.02)
```
Comment (rhoneyager) [inline src/chem/aeronet_aod2ioda.py:149]: Example for writing single-precision float data (see comment above).

Comment (bhuang95) [reply src/chem/aeronet_aod2ioda.py]: @rhoneyager Thanks, Ryan. I have updated the bit type for AOD values. Not sure why the rest three checks still failed.

**Comment by jeromebarre** [comment]
what is the status on this?

**Comment by bhuang95** [comment]
I updated the codes to address the comments a while ago. It passed Ctest. Not sure why there is still failure for three AWS check test. Anything else I need to change? Thanks.

**Comment by srherbener** [comment]
@bhuang95 the test failures are due to 2 recent PRs into the ioda repository. One PR changed the default initialization of dimension variables from 1..n to all zeros, and the other PR added default units attributes to latitude and longitude variables. These PRs account for the differences between the test output files (`build_area/iodaconv/test/testrun/files`) and the test reference files (`build_area/iodaconv/test/testoutput/files`). Because of this, it's okay to just copy the new test output files to the corresponding test reference files, and doing so should get the tests passing.

File: src/chem/aeronet_aaod2ioda.py [hunk 2]
```diff
@@ -133,83 +125,106 @@ def add_data(infile):
...
-            outdata[varDict[key]['errKey']] = np.array(nc.default_fillvals['f4'], dtype=np.float32)
-
-    # Define global atrributes
-    AttrData = {'observation_type': 'AERONET AAOD',
-                'sensor': "aeronet",
-                'surface_type': 'ocean=0, land=1, costal=2'}
+    outdata[('datetime', 'MetaData')] = d
+    varAttrs[('datetime', 'MetaData')]['units'] = ''
+
+    outdata[('frequency', 'VarMetaData')] = np.float32(frequency)
```
Comment (srherbener) [inline src/chem/aeronet_aaod2ioda.py:212]: Please place all metadata variables in the group "MetaData". We have deprecated the "VarMetaData" group. Thanks!

File: src/chem/aeronet_aod2ioda.py [hunk 3]
```diff
@@ -104,76 +97,92 @@ def add_data(infile):
...
-    AttrData = {'observation_type': 'Aod',
-                'sensor': "aeronet",
-                'surface_type': 'ocean=0,land=1,costal=2'}
+    d = np.empty([nlocs], dtype=object)
+    for i in range(nlocs):
+        d[i] = f3.time[i].strftime('%Y-%m-%dT%H:%M:%SZ')
+    outdata[('datetime', 'MetaData')] = d
+    varAttrs[('datetime', 'MetaData')]['units'] = ''
+
+    outdata[('frequency', 'VarMetaData')] = np.float32(frequency)
```
Comment (srherbener) [inline src/chem/aeronet_aod2ioda.py:170]: Please place all metadata variables in the group "MetaData". We have deprecated the "VarMetaData" group. Thanks!

File: src/chem/aeronet_aaod2ioda.py [hunk 4]
```diff
@@ -133,83 +125,106 @@ def add_data(infile):
...
+        varAttrs[key, iconv.OvalName()]['coordinates'] = 'longitude latitude station_elevation'
+        varAttrs[key, iconv.OvalName()]['units'] = 'unitless'
+        varDict[key]['errKey'] = key, iconv.OerrName()
+        varAttrs[key, iconv.OerrName()]['_FillValue'] = -999.
+        varAttrs[key, iconv.OerrName()]['coordinates'] = 'longitude latitude station_elevation'
+        varAttrs[key, iconv.OerrName()]['units'] = 'unitless'
+        varDict[key]['qcKey'] = key, iconv.OqcName()
+        varAttrs[key, iconv.OqcName()]['_FillValue'] = -999
+        varAttrs[key, iconv.OqcName()]['coordinates'] = 'longitude latitude station_elevation'
+        varAttrs[key, iconv.OqcName()]['units'] = ''
```
Comment (srherbener) [inline src/chem/aeronet_aaod2ioda.py:170]: Please check the values being assigned to the units attributes. What we finally decided on is:

* SI units when appropriate as specified in the spreadsheet (eg, air_temperature -> 'K')
* For string variables, use an empty string (ie, '')
* For unit-less variables
    * Categorical variables (eg, PreQC) use "unitless"
    * Mathematical variables (eg, mixing ratio)
        * If hidden units (eg, mixing ratio) express them - eg, "kg kg-1"
        * If pure numeric use "1"

The reason I'm asking is that it seems to me that AOD is a pure numeric quantity (unitless ratio with physical meaning) and would therefore have units set to '1'.

Thanks!

File: src/chem/aeronet_aod2ioda.py [hunk 5]
```diff
@@ -104,76 +97,92 @@ def add_data(infile):
...
+        varAttrs[key, iconv.OvalName()]['coordinates'] = 'longitude latitude station_elevation'
+        varAttrs[key, iconv.OvalName()]['units'] = 'unitless'
+        varDict[key]['errKey'] = key, iconv.OerrName()
+        varAttrs[key, iconv.OerrName()]['_FillValue'] = -999.
+        varAttrs[key, iconv.OerrName()]['units'] = 'unitless'
+        varAttrs[key, iconv.OerrName()]['coordinates'] = 'longitude latitude station_elevation'
+        varDict[key]['qcKey'] = key, iconv.OqcName()
+        varAttrs[key, iconv.OqcName()]['_FillValue'] = -999
+        varAttrs[key, iconv.OqcName()]['coordinates'] = 'longitude latitude station_elevation'
+        varAttrs[key, iconv.OqcName()]['units'] = ''
```
Comment (srherbener) [inline src/chem/aeronet_aod2ioda.py:142]: Please check the values being assigned to the units attributes. What we finally decided on is:

* SI units when appropriate as specified in the spreadsheet (eg, air_temperature -> 'K')
* For string variables, use an empty string (ie, '')
* For unit-less variables
    * Categorical variables (eg, PreQC) use "unitless"
    * Mathematical variables (eg, mixing ratio)
        * If hidden units (eg, mixing ratio) express them - eg, "kg kg-1"
        * If pure numeric use "1"

The reason I'm asking is that it seems to me that AOD is a pure numeric quantity (unitless ratio with physical meaning) and would therefore have units set to '1'.

Thanks!

**Review by srherbener — CHANGES_REQUESTED** [review]
In general the code looks great. Thanks for adding this.

My comments are concerning the setting of units and group names.

File: src/chem/aeronet_aod2ioda.py (hunk 5 above)
Comment (bhuang95) [reply src/chem/aeronet_aod2ioda.py:142]: Will modify accordingly later. Thanks.
Bo

Comment (bhuang95) [reply src/chem/aeronet_aod2ioda.py]: Steve, I have modified the codes as suggested. Seems working fine that passed all 5 checks. Please let me know if further change is needed. Thanks. 
Bo

**Review by srherbener — APPROVED** [review]
Thanks for addressing my comments! Looks good.

**Review by jeromebarre — APPROVED** [review]

**Comment by srherbener** [comment]
@CoryMartin-NOAA have all your comments been addressed? Okay to merge? Thanks!

**Review by CoryMartin-NOAA — APPROVED** [review]

**Pull request merged**